### ISSUE TITLE
Feature request and implement: Progressive download for WebP images

### ISSUE THREAD

**Issue by dreampiggy** [rs/SDWebImage#1986]
### New Issue Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | all
 Platform Version        | 7.0 above
 SDWebImage Version      | master(4.1.x)
 Integration Method      | cocoapods
 Xcode Version           | Xcode 7.3.2
 Repro rate              | e.g. all the time (100%) / sometimes x% / only once
 Repro with our demo prj | e.g. does it happen with our demo project?
 Demo project link       | e.g. link to a demo project that highlights the issue

### Issue Description and Steps

This is an feature request issue. I completed the implementation of progressive download for WebP images. But it's not suitable to create an Pull Request immediately(lack of race condition test or something). And this also include some code refactor which is on 4.1.1 milstone. So I think create a new issue about this feature and to discuss is better.

Changes:
1. Add progressive decoding using libwebp's `WebPIDecoder`. Because currently we use `ImageIO` to load progressive images. But it does not support WebP and will fail on `CGImageSourceCreateWithData`.
2. Refactor code about current progressive code after than long `if ((self.options & SDWebImageDownloaderProgressiveDownload) && self.expectedSize > 0) {xxxxxxx}`. I create two method called `progressiveDownloadForImageIO` and `progressiveDownloadForWebP` to do these procedure.

**Comment by dreampiggy** [comment]
The demo is here : 

![untitled4](https://user-images.githubusercontent.com/6919743/28997399-2acf60ce-7a46-11e7-8cc1-23472589c4e3.gif)

**Comment by dreampiggy** [comment]
Pull request is here:  #1987

**Comment by dreampiggy** [comment]
Merged via #1991.
